openhands/events/stream.py

```diff
@@ -141,7 +141,12 @@ def add_event(self, event: Event, source: EventSource):
         for key in sorted(self._subscribers.keys()):
             stack = self._subscribers[key]
             callback = stack[-1]
-            asyncio.create_task(callback(event))
+            asyncio.get_event_loop().run_in_executor(
+                None, self._callback, callback, event
+            )
+
```

Comment:
Should we also get the exception and throw them out?

I'm getting this during eval after merging this PR:
```
ERROR:asyncio:Future exception was never retrieved                                                                                [1598/1971]future: <Future finished exception=HTTPError('404 Client Error: Not Found for url: https://f067042f-17a1-4790-9dbf-44254098ebc4.eval-runtime/.
all-hands.dev/alive')>                                                                                                                       Traceback (most recent call last):
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/concurrent/futures/thread.py", line 58, in run                                          result = self.fn(*self.args, **self.kwargs)                                                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                File "/home/xingyaow/OpenHands-main-baseline/openhands/events/stream.py", line 149, in _callback
    asyncio.run(callback(event))
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)                                                                                                                             ^^^^^^^^^^^^^^^^
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/asyncio/runners.py", line 118, in run                                                   return self._loop.run_until_complete(task)                                                                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                 File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^                                                                                                                     File "/home/xingyaow/OpenHands-main-baseline/openhands/runtime/runtime.py", line 119, in on_event
    observation = self.run_action(event)                                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/OpenHands-main-baseline/openhands/runtime/remote/runtime.py", line 264, in run_action                                     self._wait_until_alive()                                                                                                                   File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 475,
 in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 376,
 in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 398,
 in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 478,
 in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/OpenHands-main-baseline/openhands/runtime/remote/runtime.py", line 216, in _wait_until_alive
    response = send_request(
               ^^^^^^^^^^^^^
  File "/home/xingyaow/OpenHands-main-baseline/openhands/runtime/utils/request.py", line 66, in send_request
    return _send_request_with_retry()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 336,
 in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 475,
 in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 376,
 in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 418,
 in exc_check
    raise retry_exc.reraise()
          ^^^^^^^^^^^^^^^^^^^
          ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 185,
 in reraise
    raise self.last_attempt.result()
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/micromamba/envs/oh/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/tenacity/__init__.py", line 478,
 in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/home/xingyaow/OpenHands-main-baseline/openhands/runtime/utils/request.py", line 63, in _send_request_with_retry
    response.raise_for_status()
  File "/home/xingyaow/.cache/pypoetry/virtualenvs/openhands-ai-YCJrcSiT-py3.12/lib/python3.12/site-packages/requests/models.py", line 1024,
in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://f067042f-17a1-4790-9dbf-44254098ebc4.eval-runtime.all-hands.dev/a
live
```

Reply:
I actually updated this again - I'm trying to not have synchronization code bleed into the lower levels. So event streams explicitly use a thread, and any open ended event from listen.py should trigger a thread.

Reply:
I think the issue got back again :(

<img width="1417" alt="image" src="https://github.com/user-attachments/assets/1ef72393-61bd-4273-a1e3-aa07be6b3d01">